pkg/square/builder.go

```diff
@@ -0,0 +1,238 @@
+		}
+	}
+
+	// begin to iteratively add blobs to the sparse share splitter calculating the actual padding
+	nonReservedStart := c.txCounter.Size() + c.pfbCounter.Size()
+	cursor := nonReservedStart
+	endOfLastBlob := nonReservedStart
+	blobWriter := shares.NewSparseShareSplitter()
+	for i, element := range c.blobs {
+		// NextShareIndex returned where the next blob should start so as to comply with the share commitment rules
+		// We fill out the remaining
+		cursor, _ = shares.NextShareIndex(cursor, element.shareSize, ss)
+		if i == 0 {
+			nonReservedStart = cursor
+		}
+
+		// defensively check that the actual padding never exceeds the max padding initially allocated for it
+		padding := cursor - endOfLastBlob
+		if padding > element.maxPadding {
+			return nil, fmt.Errorf("blob has %d padding shares, but %d was the max possible", padding, element.maxPadding)
+		}
+
+		// record the starting share index of the blob in the PFB that paid for it
+		c.pfbs[element.pfbIndex].ShareIndexes[element.blobIndex] = uint32(cursor)
+		// If this is not the first blob, we add padding by writing padded shares to the previous blob
+		// (which could be of a different namespace)
+		if i > 0 {
+			if err := blobWriter.WriteNamespacedPaddedShares(padding); err != nil {
+				return nil, fmt.Errorf("writing padding into sparse shares: %w", err)
+			}
+		}
+		// Finally write the blob itself
+		if err := blobWriter.Write(element.blob); err != nil {
+			return nil, fmt.Errorf("writing blob into sparse shares: %w", err)
+		}
+		// increment the cursor by the size of the blob
+		cursor += element.shareSize
+		endOfLastBlob = cursor
+	}
+
+	// write all the pay for blob transactions into compact shares. We need to do this after allocating the blobs to their
+	// appropriate shares as the starting index of each blob needs to be included in the PFB transaction
+	pfbTxWriter := shares.NewCompactShareSplitter(namespace.PayForBlobNamespace, appconsts.ShareVersionZero)
+	for _, iw := range c.pfbs {
+		iwBytes, err := iw.Marshal()
+		if err != nil {
+			return nil, fmt.Errorf("marshaling pay for blob tx: %w", err)
+		}
+		if err := pfbTxWriter.WriteTx(iwBytes); err != nil {
+			return nil, fmt.Errorf("writing pay for blob tx into compact shares: %w", err)
+		}
+	}
+
+	// defensively check that the counter is always greater in share count than the pfbTxWriter.
+	if c.pfbCounter.Size() < pfbTxWriter.Count() {
+		return nil, fmt.Errorf("pfbCounter.Size() < pfbTxWriter.Count(): %d < %d", c.pfbCounter.Size(), pfbTxWriter.Count())
+	}
+
+	// Write out the square
+	return WriteSquare(txWriter, pfbTxWriter, blobWriter, nonReservedStart, ss)
+}
+
+func (c *Builder) canFit(shareNum int) bool {
+	return c.currentSize+shareNum <= c.maxCapacity
+}
+
+func (c *Builder) isEmpty() bool {
+	return c.txCounter.Size() == 0 && c.pfbCounter.Size() == 0
+}
+
+type element struct {
+	blob       core.Blob
+	pfbIndex   int
+	blobIndex  int
+	shareSize  int
```

Comment:
[not blocking naming thing]
is it possible to use a different word besides shareSize that describes what this is? This is the number of shares used correct? the shares size makes me think of https://github.com/celestiaorg/celestia-app/blob/f31ee58890b1f94815057b4bbd77d7cbe62172f4/pkg/appconsts/appconsts.go#L21